src/BiDirectionalLinkedListImp.h

```diff
@@ -0,0 +1,152 @@
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::insert(const Type &newValue) {
+    if (this->first == nullptr) {
+        this->first = make_shared<BiDirectionalNode<Type>>();
+
+        this->first->setNext(nullptr);
+        this->first->setPrevious(nullptr);
+        this->first->setValue(newValue);
+
+        this->last = this->first;
+    } else {
+        shared_ptr<BiDirectionalNode<Type>> newLast = make_shared<BiDirectionalNode<Type>>();
+
+        newLast->setNext(nullptr);
+        newLast->setPrevious(last);
+        newLast->setValue(newValue);
+
+        last->setNext(newLast);
+
+        this->last = newLast;
+    }
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>> BiDirectionalLinkedList<Type>::findNodeByValue(const Type &valueToFind) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        if (valueToFind == current->getValue()) {
+            return current;
+        }
+    }
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::deleteNodeByValue(const Type &valueToFind) {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        if (valueToFind == current->getValue()) {
+            remove(current);
+
+            break;
+        }
+    }
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::printList(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->first; current != nullptr; current = current->getNext()) {
+        os << current->getValue() << " ";
+    }
+
+    os << std::endl;
+}
+
+template<class Type>
+void BiDirectionalLinkedList<Type>::printListReverse(ostream &os) const {
+    for (shared_ptr<BiDirectionalNode<Type>> current = this->last; current != nullptr;
+         current = current->getPrevious().lock()) {
+        os << current->getValue() << " ";
+    }
+
+    os << std::endl;
+}
+
+template<class Type>
+shared_ptr<BiDirectionalNode<Type>>
+BiDirectionalLinkedList<Type>::insertNodeAfter(shared_ptr<BiDirectionalNode<Type>> nodeToInsertAfter,
+                                               shared_ptr<BiDirectionalNode<Type>> newNode) {
+    // If list is empty we insert it to the the head of the list
+    if (getFirst() == nullptr) {
+        assert(nodeToInsertAfter == nullptr);
+
+        this->first = newNode;
+        this->last = newNode;
+
+        return getFirst();
+    }
+
+    newNode->setPrevious(nodeToInsertAfter);
+
+    if (nodeToInsertAfter != nullptr) {
+        newNode->setNext(nodeToInsertAfter->getNext());
+        nodeToInsertAfter->setNext(newNode);
+    } else {
+        newNode->setNext(this->first);
+        this->first = newNode;
+    }
+
+    if (newNode->getNext() != nullptr) {
+        newNode->getNext()->setPrevious(newNode);
+    }
+
+    if (nodeToInsertAfter == getLast()) {
+        this->last = newNode;
+    }
+
+    return newNode;
+}
```

Comment:
ok this entire function seems suspicious 🤨

Reply:
Well you need to point out something because it's necessary for efficiency.